lua/plenary/path.lua

```diff
@@ -314,6 +314,50 @@ function Path:rmdir()
   uv.fs_rmdir(self:absolute())
 end
 
+function Path:rename(opts)
+  opts = opts or {}
+  if not opts.new_name or opts.new_name == "" then
+    error("Please provide the new name!")
+  end
+
+    -- handles `.`, `..`, `./`, and `../`
+  if opts.new_name:match('^%.%.?/?\\?.+') then
+    opts.new_name = {
+      uv.fs_realpath(opts.new_name:sub(1, 3)),
+      opts.new_name:sub(4, #opts.new_name)
+    }
+  end
+
+  local new_path = Path:new(opts.new_name)
+
+  if new_path:exists() then
+    error('File or directory already exists!')
+  end
+
+  local success = uv.fs_rename(self:absolute(), new_path:absolute(), function(err)
+    if err then
+      error(err)
+    end
+  end)
+
+  local path_tbl = vim.split(new_path.filename, path.sep)
+  self.filename = path_tbl[#path_tbl]
```

Comment:
Are we only interested in the last part? Is that correct. Will `self:absoule` gives us the correct path afterwards? I gonna try it but i don't think so.

Reply:
yeah, that gives the correct path when we move it to a different directory. But now I don't know how to test that for github action. I can hardcode the path locally, but won't be the case with github action

Reply:
I couldn't confirm that. `self:absolute()` will still give me the old path.
Thats the script i just tested. The rename and everything so :+1: but i think `p` should point to `/home/conni/test_file.txt` afterwards
```lua
local Path = require'plenary.path'

local p = Path:new('test_file.txt')
p:touch()
print(p:absolute())
p:rename({ new_name = '/home/conni/test_file.txt' })
print(p:absolute(), p.filename)
```

`p.filename` will be `test_file.txt` but `absolute` will produce a absolute path based on `self._cwd`(still plenary dir here) and `filename` unless the path is already absolute. `cwd` will most of the time point to neovim cwd so i wouldn't change that but maybe produce a new absolute filename with `vim.loop.fs_realpath`

Reply:
wait, I think I've fixed it. Let me push it. This is the output of that snippet.

```
/home/elianiva/repos/plenary.nvim/test_file.txt
/home/elianiva/test_file.txt /home/elianiva/test_file.txt
```

Though I'm not fully understand how `uv.fs_realpath` work :rofl: It's giving me `nil` with no error sometimes and I don't understand why. Tried to search on libuv doc and luv but I didn't find anything

Reply:
`man 3 realpath`
It doesn't expand, it make this path: `/home/conni/repos/../file` into `/home/conni/file` and it can resolve links.

Reply:
Oh i just realized i forgot to mention that it will throw an error when it has to deal with something like this: `~` or `$HOME`. It will not expand paths

Reply:
right, we also need to handle that

Reply:
You can vim.fn.expand first and than `vim.loop.fs_realpath` but i am not sure right now. Tami also has written `path:expand` but i think this doesn't work as expected and doesn't handle window.

I can look at the CI weirdness later :)